lib/madwf_ml.cpp

```diff
@@ -0,0 +1,348 @@
+        // theta
+        blas::axpy(1.0, theta, tmp);
+        if (tune_suppressor) { blas::axpy(pmu, *phi, tmp); }
+
+        ref(theta, tmp);
+
+        // lambda = D' * A * D * phi
+        madwf_ml::transfer_5d_hh(tmp, ADphi, P, true);
+
+        ref(lambda, tmp);
+
+        std::vector<ColorSpinorField *> lhs {&chi, &theta, &lambda};
+        std::vector<ColorSpinorField *> rhs {&chi, &theta, &lambda};
+        Complex dot[9];
+        blas::cDotProduct(dot, lhs, rhs);
+
+        a[0] += dot[0].real();
+        a[1] += -2.0 * dot[1].real();
+        a[2] += dot[4].real() + 2.0 * dot[2].real();
+        a[3] += -2.0 * dot[5].real();
+        a[4] += dot[8].real();
+      }
+
+      std::array<double, 4> coeffs = {4.0 * a[4], 3.0 * a[3], 2.0 * a[2], a[1]};
+      auto rs = cubic_formula(coeffs);
+
+      alpha = 0;
+      double root_min = poly4(a, 0);
+      for (auto r : rs) {
+        double eval = poly4(a, r);
+        if (root_min > eval) {
+          root_min = eval;
+          alpha = r;
+        }
+      }
+
+      axpby(device_param, 0.0f, device_param, -alpha, P);
+      if (tune_suppressor) { mu -= alpha * pmu; }
+
+      if (getVerbosity() >= QUDA_SUMMARIZE) {
+        printfQuda("grad min iter %05d: %04d chi2 = %8.4e, chi2 %% = %8.4e, alpha = %+8.4e, mu = %+8.4e\n", comm_rank(),
+                   iteration, chi2, chi2 / residual, alpha, mu);
+      }
+    }
+
+    trained = true;
+
+    if (getVerbosity() >= QUDA_VERBOSE) { printfQuda("Training finished ...\n"); }
+    count = 0;
+    for (auto phi: B) {
+      double ind_chi2 = cost(ref, base, chi, *phi);
+      double phi2 = blas::norm2(*phi);
+      if (getVerbosity() >= QUDA_VERBOSE) {
+        printfQuda("chi2 %03d %% = %8.4e, phi2 = %8.4e\n", count, ind_chi2 / phi2, phi2);
+      }
+      count++;
+    }
+
+    for (auto pb: B) { delete pb; }
+
+    // Broadcast the trained parameters
+    host_param = device_param.to_host();
+    comm_broadcast(host_param.data(), host_param.size() * sizeof(transfer_float));
+    device_param.from_host(host_param);
+
+    commGlobalReductionPop();
+    profile.TPSTOP(QUDA_PROFILE_TRAINING);
+
+    if (param.madwf_param_save) {
+      profile.TPSTART(QUDA_PROFILE_IO);
+      if (comm_rank() == 0) { save_parameter(Ls, Ls_base); } // Only rank zero write out to the disk
+      comm_barrier();
+      profile.TPSTOP(QUDA_PROFILE_IO);
+    }
+
+  }
+
+  void MadwfAcc::save_parameter(int Ls, int Ls_base) {
+    std::vector<transfer_float> host_param = device_param.to_host();
+
+    std::string save_param_path(param.madwf_param_outfile);
+    char cstring[512];
+    sprintf(cstring, "/madwf_trained_param_ls_%02d_%02d_mu_%.3f.dat", Ls, Ls_base, mu);
+    save_param_path += std::string(cstring);
+    FILE *fp = fopen(save_param_path.c_str(), "w");
+    if (!fp) { errorQuda("Unable to open file %s\n", save_param_path.c_str()); }
+    size_t fwrite_count = fwrite(host_param.data(), sizeof(transfer_float), host_param.size(), fp);
```

Comment:
I'm a little confused---is every rank now writing to the same exact file? Now that the filename is independent of the rank? That seems like trouble...

Reply:
Only rank 0 will call the `save_parameter` function.

Reply:
I think it'd be safer to put the check within `save_parameter` as opposed to externally, or at least put a check inside the function that it's being called by rank 0.

Reply:
https://github.com/lattice/quda/pull/1061/commits/5cce447990c8ea085125244fea58154e4271dc8a